Feat/allow passing answer from @on to @after hook

OSkrk opened the pull request:
This PR addresses the feature discussed [here](https://github.com/mobilityhouse/ocpp/issues/264), allowing the response sent inside the @on to be passed to the @after hook for further processing (avoiding the use of temp variables). The feature enables the @after decorator to accept an `inject_response` parameter, which can be used to control whether the response from the `_on_action` handler is passed to the `_after_action` handler, and guarantees backwards compatibility.


@OrangeTux , @tropxy, @HugoJP1, @mdwcrft would like to know you guy's feedback on this.

OSkrk committed:
fix typo CostUpdated enum for 201

OSkrk committed:
Merge branch 'mobilityhouse:master' into master

OSkrk committed:
feat/allow_passing_answer_from_@on_to_@after_hook

OSkrk commented on tests/test_charge_point.py:
```diff
@@ -472,3 +472,33 @@ def after_boot_notification(self, *args, **kwargs):
+
+        @after(Action.BootNotification, inject_response=True)
+        def after_boot_notification(self, on_response, **kwargs):
+            # print args and kwargs to check if response is injected
```
```suggestion
```

OSkrk committed:
Update tests/test_charge_point.py

astrand commented on ocpp/charge_point.py:
```diff
@@ -356,10 +356,25 @@ async def _handle_call(self, msg):
+                        **snake_case_payload, on_response=response_payload
+                    )
+                else:
+                    response = handler(**snake_case_payload)
```
Not so "elegant" to have 4 calls to handler(). My suggestion is to use keyword arguments instead, like I did in https://github.com/mobilityhouse/ocpp/issues/264#issuecomment-1910237896 .

OSkrk replied:
This approach was suggested by @tropxy (one of the authors of this repo) in this [comment](https://github.com/mobilityhouse/ocpp/issues/264#issuecomment-964425685)

tropxy replied:
@OSkrk what @astrand is suggesting is something like this (right @astrand ?):

```python
if "call_unique_id" in handler_signature.parameters:
    snake_case_payload["call_unique_id"] = msg.unique_id
if  inject_response:
    snake_case_payload["on_response"] = response_payload

response = handler(**snake_case_payload)
```

May you try that and see if it works?

Note: I have used your current nomenclature, but if you change the code then "on_response" will be "call_response"

OSkrk replied:
Addressed

astrand commented on ocpp/charge_point.py:
```diff
@@ -356,10 +356,25 @@ async def _handle_call(self, msg):
             call_unique_id_required = "call_unique_id" in handler_signature.parameters
             # call_unique_id should be passed as kwarg only if is defined explicitly
             # in the handler signature
+            inject_response = getattr(handler, "_inject_response", False)
```
Why check attribute on the handler, instead of checking signature parameters? This opens up for some corner cases, like if inject_response is True but after handler does not accept "on_response" kwarg. Also it is different from how call_unique_id is handled.

OSkrk replied:
@astrand , thanks for your comment. Why would you set the flag to `True` to pass the reponse and do not accept it in the @after ?

OSkrk replied:
Keeping the explicit inject_response flag as discussed (and per @tropxy / @OrangeTux's validation). Injecting call_response is consistent with how payload kwargs are already passed... you opt in with the flag and accept it in the handler (or via **kwargs). If someone sets the flag but doesn't accept the arg, they get an immediate, clear TypeError, so it fails fast and only on explicit opt-in.

astrand reviewed (changes requested):
Thanks for working on this!

tropxy commented on ocpp/routing.py:
```diff
@@ -56,7 +56,7 @@ def inner(*args, **kwargs):
 
 
-def after(action):
+def after(action, inject_response=False):
```
can you include notes in the doc strings in respect to the new argument and its purpose?

OSkrk replied:
added a docstring section to the after() decorator in ocpp/routing.py

tropxy commented on ocpp/charge_point.py:
```diff
@@ -356,10 +356,25 @@ async def _handle_call(self, msg):
+                    response = handler(
+                        **snake_case_payload,
+                        call_unique_id=msg.unique_id,
+                        on_response=response_payload,
```
It would be better the term "call_response". The "on" prefix is given to functions/callbacks as an action will be performed via that callback function. Here we want to pass an argument..

OSkrk replied:
Addressed. The rename from on_response → call_response is done everywhere

OrangeTux commented:
I don't have anything against this PR. I leave it up to the new maintainers to approve and merge it.

I'll have a call with the new maintainers this Friday and I'll bring this PR to the table.

mdwcrft committed:
Merge branch 'master' into feat/allow_passing_answer_from_@on_to_@after_hook

OSkrk committed:
Address review: single handler() call, rename on_response->call_response, fix enum

  - Refactor _handle_call to build kwargs dict and call handler() once (astrand/tropxy)
  - Rename injected kwarg on_response -> call_response (tropxy)
  - Document inject_response argument in after() docstring (tropxy)
  - Fix test to use snake_case Action.boot_notification (fixes failing CI)

OSkrk committed:
Merge branch 'master' into feat/allow_passing_answer_from_@on_to_@after_hook

mdwcrft committed:
Merge branch 'master' into feat/allow_passing_answer_from_@on_to_@after_hook
